// mout/lang.js
"use strict";
// automatically generated, do not edit!
// run `node build` instead
exports.__esModule = true;
var GLOBAL_1 = require("./lang/GLOBAL");
exports.GLOBAL = GLOBAL_1["default"];
var clone_1 = require("./lang/clone");
exports.clone = clone_1["default"];
var createObject_1 = require("./lang/createObject");
exports.createObject = createObject_1["default"];
var ctorApply_1 = require("./lang/ctorApply");
exports.ctorApply = ctorApply_1["default"];
var deepClone_1 = require("./lang/deepClone");
exports.deepClone = deepClone_1["default"];
var deepEquals_1 = require("./lang/deepEquals");
exports.deepEquals = deepEquals_1["default"];
var defaults_1 = require("./lang/defaults");
exports.defaults = defaults_1["default"];
var inheritPrototype_1 = require("./lang/inheritPrototype");
exports.inheritPrototype = inheritPrototype_1["default"];
var is_1 = require("./lang/is");
exports.is = is_1["default"];
var isArguments_1 = require("./lang/isArguments");
exports.isArguments = isArguments_1["default"];
var isArray_1 = require("./lang/isArray");
exports.isArray = isArray_1["default"];
var isBoolean_1 = require("./lang/isBoolean");
exports.isBoolean = isBoolean_1["default"];
var isDate_1 = require("./lang/isDate");
exports.isDate = isDate_1["default"];
var isEmpty_1 = require("./lang/isEmpty");
exports.isEmpty = isEmpty_1["default"];
var isFinite_1 = require("./lang/isFinite");
exports.isFinite = isFinite_1["default"];
var isFunction_1 = require("./lang/isFunction");
exports.isFunction = isFunction_1["default"];
var isInteger_1 = require("./lang/isInteger");
exports.isInteger = isInteger_1["default"];
var isKind_1 = require("./lang/isKind");
exports.isKind = isKind_1["default"];
var isNaN_1 = require("./lang/isNaN");
exports.isNaN = isNaN_1["default"];
var isNull_1 = require("./lang/isNull");
exports.isNull = isNull_1["default"];
var isNumber_1 = require("./lang/isNumber");
exports.isNumber = isNumber_1["default"];
var isObject_1 = require("./lang/isObject");
exports.isObject = isObject_1["default"];
var isPlainObject_1 = require("./lang/isPlainObject");
exports.isPlainObject = isPlainObject_1["default"];
var isPrimitive_1 = require("./lang/isPrimitive");
exports.isPrimitive = isPrimitive_1["default"];
var isRegExp_1 = require("./lang/isRegExp");
exports.isRegExp = isRegExp_1["default"];
var isString_1 = require("./lang/isString");
exports.isString = isString_1["default"];
var isUndefined_1 = require("./lang/isUndefined");
exports.isUndefined = isUndefined_1["default"];
var isnt_1 = require("./lang/isnt");
exports.isnt = isnt_1["default"];
var kindOf_1 = require("./lang/kindOf");
exports.kindOf = kindOf_1["default"];
var toArray_1 = require("./lang/toArray");
exports.toArray = toArray_1["default"];
var toNumber_1 = require("./lang/toNumber");
exports.toNumber = toNumber_1["default"];
var toString_1 = require("./lang/toString");
exports.toString = toString_1["default"];
