// mout/math.js
"use strict";
// automatically generated, do not edit!
// run `node build` instead
exports.__esModule = true;
var ceil_1 = require("./math/ceil");
exports.ceil = ceil_1["default"];
var clamp_1 = require("./math/clamp");
exports.clamp = clamp_1["default"];
var countSteps_1 = require("./math/countSteps");
exports.countSteps = countSteps_1["default"];
var floor_1 = require("./math/floor");
exports.floor = floor_1["default"];
var inRange_1 = require("./math/inRange");
exports.inRange = inRange_1["default"];
var isNear_1 = require("./math/isNear");
exports.isNear = isNear_1["default"];
var lerp_1 = require("./math/lerp");
exports.lerp = lerp_1["default"];
var loop_1 = require("./math/loop");
exports.loop = loop_1["default"];
var map_1 = require("./math/map");
exports.map = map_1["default"];
var norm_1 = require("./math/norm");
exports.norm = norm_1["default"];
var overflow_1 = require("./math/overflow");
exports.overflow = overflow_1["default"];
var round_1 = require("./math/round");
exports.round = round_1["default"];
